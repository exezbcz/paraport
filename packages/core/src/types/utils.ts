/**
 * Extracts the union of the values of an object type.
 */
export type ObjectValues<T> = T[keyof T]
