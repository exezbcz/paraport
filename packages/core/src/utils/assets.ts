/**
 * Asset utilities built on top of the Paraspell assets catalog.
 *
 * - Look up chain-specific asset metadata (native and foreign)
 * - Produce a Paraspell `TCurrencyInput` for builder APIs (location/alias/id/symbol)
 * - Read existential deposits when provided by the catalog
 * - Check whether an asset is supported for a given origin→destination pair
 */
import type { Asset, Chain } from '@paraport/static'
import {
	ForeignAbstract,
	type TCurrencyInput,
	type TForeignAssetInfo,
	getAssetsObject,
	getSupportedAssets,
} from '@paraspell/sdk'

/**
 * Chain→asset→alias overrides used to disambiguate assets with multiple
 * catalog entries. For example, on AssetHubKusama, DOT is represented as
 * `DOT2` in Paraspell's catalog.
 */
const aliasMap: Partial<Record<Chain, Partial<Record<Asset, string>>>> = {
	AssetHubKusama: {
		DOT: 'DOT2',
	},
}

/**
 * Returns asset metadata (native or foreign) from Paraspell's catalog.
 *
 * - Searches both `nativeAssets` and `otherAssets` for the first match.
 * - When `autoalias` is true (default), additionally requires the entry's
 *   `alias` to match the chain-specific override in `aliasMap` (when present).
 *
 * @param chain - Chain identifier
 * @param symbol - Asset symbol to look up (e.g., 'DOT', 'KSM')
 * @param options.autoalias - Enforce alias matching when configured
 * @returns Matching asset info or undefined when not found
 */
export const getAssetInfo = (
	chain: Chain,
	symbol: Asset,
	{ autoalias = true } = {},
) => {
	const { nativeAssets, otherAssets } = getAssetsObject(chain)
	const alias = aliasMap[chain]?.[symbol]

	return [...nativeAssets, ...otherAssets].find((asset) =>
		autoalias
			? asset.symbol === symbol && asset.alias === alias
			: asset.symbol === symbol,
	)
}

/**
 * Looks up asset metadata and throws when the asset is not found on `chain`.
 *
 * @param chain - Chain identifier
 * @param symbol - Asset symbol to look up
 * @param options.autoalias - Enforce alias matching when configured
 * @throws Error when the asset cannot be found for the chain
 * @returns Matching asset info
 */
export const getAssetInfoOrThrow = (
	chain: Chain,
	symbol: Asset,
	{ autoalias = true } = {},
) => {
	const assetInfo = getAssetInfo(chain, symbol, { autoalias })

	if (!assetInfo) {
		throw new Error(`Asset ${symbol} not found for ${chain}`)
	}

	return assetInfo
}

/**
 * Translates an asset on `chain` to a Paraspell `TCurrencyInput` discriminator.
 *
 * Order of precedence:
 * - If the asset has a `location`, return `{ location }` (ForeignAssets/XCM)
 * - Else if it has an `alias`, return `{ symbol: ForeignAbstract(alias) }`
 * - Else if it has an `assetId`, return `{ id }` (Assets pallet)
 * - Else fall back to `{ symbol }` (native or symbol-only identification)
 *
 * @param chain - Chain identifier
 * @param symbol - Asset symbol
 * @returns Paraspell-compatible currency input discriminator
 * @throws Error when the asset is not present in the catalog for the chain
 */
export const getParaspellCurrencyInput = (
	chain: Chain,
	symbol: Asset,
): TCurrencyInput => {
	const assetInfo = getAssetInfoOrThrow(chain, symbol)

	if (assetInfo.location) {
		return { location: assetInfo.location }
	}

	if (assetInfo.alias) {
		return { symbol: ForeignAbstract(assetInfo.alias) }
	}

	const assetId = (assetInfo as TForeignAssetInfo).assetId

	if (assetId) {
		return { id: assetId }
	}

	return { symbol: assetInfo.symbol }
}

/**
 * Resolves the existential deposit for a given asset on `chain`, when
 * available in the catalog entry.
 *
 * @param chain - Chain identifier
 * @param symbol - Asset symbol
 * @returns Existential deposit value if present; otherwise undefined
 */
export const getAssetExistentialDeposit = (chain: Chain, symbol: Asset) => {
	const asset = getAssetInfo(chain, symbol)

	return asset?.existentialDeposit
}

/**
 * Checks whether `asset` is supported from `origin` to `destination` using
 * Paraspell's `getSupportedAssets` list for the pair.
 *
 * @param origin - Origin chain
 * @param destination - Destination chain
 * @param asset - Asset symbol
 * @returns True when supported; false otherwise
 */
export const isAssetSupported = (
	origin: Chain,
	destination: Chain,
	asset: Asset,
): boolean => {
	const originInfo = getAssetInfo(origin, asset)
	const supported = getSupportedAssets(origin, destination)

	if (originInfo?.alias) {
		return supported.some(
			(info: any) => info.symbol === asset && info.alias === originInfo.alias,
		)
	}

	return supported.some((info: any) => info.symbol === asset)
}
