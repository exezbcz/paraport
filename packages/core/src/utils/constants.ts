import { Asset, Chain, type ChainPrefix } from '../types'

export const ASSET_CHAINS_MAP: Partial<Record<Asset, Chain[]>> = {
	[Asset.DOT]: [Chain.POLKADOT, Chain.ASSETHUBPOLKADOT],
	[Asset.KSM]: [Chain.KUSAMA, Chain.ASSETHUBKUSAMA],
}

export const CHAIN_PREFIX_TO_CHAIN_MAP: Partial<Record<ChainPrefix, Chain>> = {
	dot: Chain.POLKADOT,
	ahp: Chain.ASSETHUBPOLKADOT,
	ksm: Chain.KUSAMA,
	ahk: Chain.ASSETHUBKUSAMA,
}

export const CHAIN_TO_CHAIN_PREFIX_MAP: Partial<Record<Chain, ChainPrefix>> = {
	[Chain.POLKADOT]: 'dot',
	[Chain.ASSETHUBPOLKADOT]: 'ahp',
	[Chain.KUSAMA]: 'ksm',
	[Chain.ASSETHUBKUSAMA]: 'ahk',
}
