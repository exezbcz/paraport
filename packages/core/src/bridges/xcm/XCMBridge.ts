import { Initializable } from '@/base/Initializable'
import type BalanceService from '@/services/BalanceService'
import type { Balance } from '@/services/BalanceService'
import type PolkadotApi from '@/services/PolkadotApi'
import type {
	BridgeAdapter,
	BridgeProtocol,
	BridgeTransferParams,
} from '@/types/bridges'
import type { Quote, SDKConfig } from '@/types/common'
import {
	type TeleportMode,
	TeleportModes,
	type TeleportParams,
} from '@/types/teleport'
import type { TransactionCallback } from '@/types/transactions'
import { formatAddress, getRouteChains, isRouteDisabled } from '@/utils'
import { getParaspellCurrencyInput } from '@/utils/assets'
import { signAndSend } from '@/utils/tx'
import type { Asset, Chain } from '@paraport/static'
import { Builder } from '@paraspell/sdk'
import { maxBy } from 'lodash'

type XCMTransferParams = {
	amount: bigint
	originChain: Chain
	destinationChain: Chain
	address: string
	asset: Asset
}

/**
 * XCM bridge adapter using Paraspell to construct and submit XCM transactions.
 */
export default class XCMBridge extends Initializable implements BridgeAdapter {
	protocol: BridgeProtocol = 'XCM'
	private readonly requiredSignatureCount: number = 1

	constructor(
		private readonly config: SDKConfig,
		private readonly balanceService: BalanceService,
		private readonly papi: PolkadotApi,
	) {
		super()
	}

	/**
	 * Builds a typed Paraspell query builder for XCM transfers.
	 * @param params - XCM transfer parameters
	 * @returns Paraspell builder instance
	 */
	private getParaspellQuery({
		amount,
		originChain,
		destinationChain,
		address,
		asset,
	}: XCMTransferParams) {
		const currencyInput = getParaspellCurrencyInput(originChain, asset)

		return (
			Builder({
				// Provide chain-specific clients to avoid paraspell from instantiating its own papi instances
				apiOverrides: {
					[originChain]: this.papi.getInstance(originChain).client,
					[destinationChain]: this.papi.getInstance(destinationChain).client,
				},
			})
				.from(originChain)
				.to(destinationChain)
				.currency({ ...currencyInput, amount })
				.address(formatAddress(address, destinationChain))
				.senderAddress(formatAddress(address, originChain))
				// Pay transaction fee with the same asset
				.feeAsset(currencyInput)
		)
	}

	/**
	 * Estimates the XCM fees for origin and destination.
	 * @param params - XCM transfer parameters
	 * @returns Sum of origin and destination fees as bigint
	 */
	private async getXcmFee({
		amount,
		originChain,
		destinationChain,
		address,
		asset,
	}: XCMTransferParams): Promise<bigint> {
		const query = this.getParaspellQuery({
			amount,
			originChain,
			destinationChain,
			address,
			asset,
		})

		try {
			const { origin, destination } = await query.getXcmFee()

			return BigInt(origin.fee || 0) + BigInt(destination.fee || 0)
		} catch (error) {
			console.log('Failed getting Xcm fee', error)
			throw error
		}
	}

	/**
	 * Computes transfer amount based on teleport mode and fees.
	 * @param amount - User requested amount
	 * @param xcmFee - Estimated XCM fee
	 * @param currentChainBalance - Balance on target chain
	 * @param teleportMode - Teleport mode
	 * @returns Calculated send amount
	 */
	private calculateTeleportAmount({
		amount,
		xcmFee,
		currentChainBalance,
		teleportMode,
	}: {
		amount: bigint
		teleportMode: TeleportMode
		currentChainBalance: Balance
		xcmFee: bigint
	}) {
		if (TeleportModes.Expected === teleportMode) {
			return amount - currentChainBalance.transferable + xcmFee
		}
		if (TeleportModes.Only === teleportMode) {
			return amount - xcmFee
		}
		return amount // TeleportModes.Exact
	}

	/**
	 * Produces a quote for teleporting funds via XCM given input params.
	 *
	 * @param params - Teleport parameters
	 * @returns Quote or null if not feasible
	 */
	async getQuote({
		address,
		asset,
		chain: destinationChain,
		amount,
		teleportMode = TeleportModes.Expected,
	}: TeleportParams): Promise<Quote | null> {
		// 1. get chains where the token is available
		const chains = getRouteChains(destinationChain, asset).filter((chain) =>
			this.config.chains.includes(chain),
		)

		// 2. get address balances on all chains where the token is available
		const balances = await this.balanceService.getBalances({
			address,
			chains,
			asset,
		})

		const currentChainBalance = balances.find(
			(balance) => balance.chain === destinationChain,
		)

		// 3. from possible target chains find the one with the highest transferable balance
		const targetChainBalances = balances.filter(
			(balance) => balance.chain !== destinationChain,
		)

		const highestBalanceChain = maxBy(
			targetChainBalances,
			(balance) => balance.transferable,
		)

		if (!highestBalanceChain || !currentChainBalance) {
			return null
		}

		const originChain = highestBalanceChain.chain

		// 4. calculate amount to teleport
		const simpleXcmFee = await this.getXcmFee({
			amount,
			originChain,
			destinationChain,
			address,
			asset,
		})

		const teleportAmountWithXcmFee = this.calculateTeleportAmount({
			amount,
			xcmFee: simpleXcmFee,
			currentChainBalance: currentChainBalance,
			teleportMode: TeleportModes.Expected,
		})

		const xcmFee = await this.getXcmFee({
			amount: teleportAmountWithXcmFee,
			originChain,
			destinationChain,
			address,
			asset,
		})

		const transferAmount = this.calculateTeleportAmount({
			amount,
			xcmFee,
			currentChainBalance,
			teleportMode,
		})

		const query = this.getParaspellQuery({
			amount: transferAmount,
			originChain,
			destinationChain,
			address,
			asset,
		})

		const dryRun = await query.dryRun()

		const totalFees = xcmFee
		const receivingAmount = transferAmount - totalFees

		if (
			transferAmount <= 0n || // valid transfer amount
			receivingAmount <= 0n || // ensures positive receive
			Boolean(dryRun.failureReason) || // fails to execute
			highestBalanceChain.transferable < transferAmount || // can transfer
			(teleportMode === TeleportModes.Expected &&
				currentChainBalance.transferable + receivingAmount < amount) // ends up with desired amount
		) {
			return null
		}

		return {
			teleportMode,
			amount: receivingAmount,
			total: transferAmount,
			asset: asset,
			route: {
				origin: originChain,
				destination: destinationChain,
				protocol: this.protocol,
			},
			fees: {
				bridge: xcmFee,
				total: totalFees,
			},
			execution: {
				requiredSignatureCount: this.requiredSignatureCount,
				timeMs: 30000,
			},
		}
	}

	/**
	 * Executes an XCM transfer and wires events to a transaction callback.
	 *
	 * @param params - Transfer parameters
	 * @param callback - Receives transaction lifecycle updates
	 * @returns Unsubscribe function for the underlying transaction observer
	 */
	async transfer(
		{
			amount,
			from: originChain,
			to: destinationChain,
			address,
			asset,
		}: BridgeTransferParams,
		callback: TransactionCallback,
	) {
		if (isRouteDisabled(originChain, destinationChain)) {
			throw new Error(
				`XCM route between ${originChain}>${destinationChain} temporarily disabled`,
			)
		}

		const query = this.getParaspellQuery({
			amount,
			originChain,
			destinationChain,
			address,
			asset,
		})

		const transaction = await query.build()

		return signAndSend({
			transaction,
			callback,
			signer: await this.config.getSigner(),
		})
	}

	/**
	 * Initializes the adapter. Pre-warming of APIs could be added here.
	 * @returns Promise resolving when ready
	 */
	async initialize(): Promise<void> {
		return Promise.resolve()
	}
}
