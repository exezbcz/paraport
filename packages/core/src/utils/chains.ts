import {
	type Asset,
	CHAINS,
	CHAIN_NAMES,
	type Chain,
	type ChainProperties,
	Chains,
} from '@paraport/static'
import { SUBSTRATE_CHAINS } from '@paraspell/sdk'
import { isAssetSupported } from './assets'

/**
 * Returns static chain properties (ss58, decimals, explorer, etc.).
 * @param chain - Chain identifier
 * @returns Chain properties
 */
export const chainPropListOf = (chain: Chain): ChainProperties => {
	return CHAINS[chain]
}

/**
 * Gets the SS58 address format for a chain.
 * @param chain - Chain identifier
 * @returns ss58 format number
 */
export const ss58Of = (chain: Chain): number => {
	return chainPropListOf(chain).ss58Format
}

/**
 * Gets token decimals for a chain.
 * @param chain - Chain identifier
 * @returns Number of decimals
 */
export const decimalsOf = (chain: Chain): number => {
	return chainPropListOf(chain).tokenDecimals
}

/**
 * Lists chains where teleport will interact and where a given asset is available.
 *
 * Returns the destination `chain` and all origin chains that support the asset.
 *
 * @param chain Destination chain.
 * @param asset Asset symbol.
 * @returns Array of chains including the destination and eligible origins.
 */
export const getRouteChains = (chain: Chain, asset: Asset): Chain[] => {
	const otherChains = SUBSTRATE_CHAINS.filter((subChain) => {
		if (chain === subChain) return false

		return isAssetSupported(chain, subChain as Chain, asset)
	})

	const allowed = new Set(Object.values(Chains))

	return [chain, ...otherChains].filter((c) =>
		allowed.has(c as Chain),
	) as Chain[]
}

/**
 * Gets a human-readable chain name.
 * @param chain - Chain identifier
 * @returns Chain name
 */
export const getChainName = (chain: Chain): string => {
	return CHAIN_NAMES[chain]
}

/**
 * Returns the block explorer URL for a chain.
 * @param chain - Chain identifier
 * @returns Explorer base URL
 */
export const blockExplorerOf = (chain: Chain): string => {
	return chainPropListOf(chain).blockExplorer
}
