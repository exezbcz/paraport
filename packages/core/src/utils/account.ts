import { ss58Of } from '@/utils/chains'
import type { Chain } from '@paraport/static'
import { decodeAddress, encodeAddress } from 'dedot/utils'

/**
 * Re-encodes an address to the target chain's ss58 format.
 * @param address - Address string (any valid ss58 prefix)
 * @param chain - Target chain
 * @returns Address encoded for the target chain
 */
export const formatAddress = (address: string, chain: Chain) =>
	encodeAddress(address, ss58Of(chain))

/**
 * Validates if a string is a valid ss58 address.
 * @param address - Address string
 * @returns True when the address can be decoded/encoded
 */
export const isValidAddress = (address: string) => {
	try {
		encodeAddress(decodeAddress(address))
		return true
	} catch (error) {
		return false
	}
}
