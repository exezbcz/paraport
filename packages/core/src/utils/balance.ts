import type { Chain } from '../types/common'
import { edOf } from './chains'

export const transferableBalanceOf = (amount: string, chain: Chain): number => {
	return Number(amount) - edOf(chain)
}
