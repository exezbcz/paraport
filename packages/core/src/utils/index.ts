import { formatAddress } from '@/utils/account'
import { transferableBalanceOf } from '@/utils/balance'
import { edOf, teleportEdOf } from '@/utils/chains'
import {
	ASSET_CHAINS_MAP,
	type Asset,
	CHAINS,
	type Chain,
	type ChainProperties,
	ENDPOINT_MAP,
} from '@paraport/static'

export const chainPropListOf = (chain: Chain): ChainProperties => {
	return CHAINS[chain]
}

export const ss58Of = (chain: Chain): number => {
	return chainPropListOf(chain).ss58Format
}

export const decimalsOf = (chain: Chain): number => {
	return chainPropListOf(chain).tokenDecimals
}

export const getChainsOfAsset = (asset: Asset): Chain[] => {
	return ASSET_CHAINS_MAP[asset] || []
}

export const endpointOf = (chain: Chain): string => {
	return ENDPOINT_MAP[chain]
}

export { formatAddress, edOf, teleportEdOf, transferableBalanceOf }
