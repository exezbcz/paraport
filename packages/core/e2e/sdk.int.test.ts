import { SUBSTRATE_ADDRESS } from '@/__tests__/utils/constants'
import XCMBridge from '@/bridges/xcm/XCMBridge'
import { Assets, Chains } from '@paraport/static'
import { describe, expect, it, vi } from 'vitest'

// Make the suite single-threaded to avoid any accidental shared state
describe.sequential(
	'Integration: XCMBridge transfer (serialized tx snapshot)',
	() => {
		// Mock utils used by the bridge (no formatting side-effects, deterministic)
    vi.mock('@/utils', () => ({
        formatAddress: vi.fn((a: string) => a),
        getChainsOfAsset: vi.fn(() => [Chains.Kusama, Chains.AssetHubKusama]),
        getRouteChains: vi.fn((_dest: string, _asset: string) => [Chains.Kusama, Chains.AssetHubKusama]),
    }))

		// Use paraspell Builder mock with fixed fees and a serializable tx
		vi.mock('@paraspell/sdk', () => {
			const builderChain = {
				from: vi.fn().mockReturnThis(),
				to: vi.fn().mockReturnThis(),
				currency: vi.fn().mockReturnThis(),
				address: vi.fn().mockReturnThis(),
				senderAddress: vi.fn().mockReturnThis(),
				feeAsset: vi.fn().mockReturnThis(),
				dryRun: vi.fn().mockResolvedValue({}),
				getXcmFee: vi.fn().mockResolvedValue({
					origin: { fee: 10n },
					destination: { fee: 5n },
				}),
				build: vi.fn().mockResolvedValue({
					decodedCall: {
						module: 'PolkadotXcm',
						method: 'transfer_assets',
						parameters: {
							assets: '...omitted for brevity...',
							dest: '...omitted for brevity...',
						},
					},
					signSubmitAndWatch: vi.fn(),
				}),
			}
			return {
				Builder: vi.fn().mockReturnValue(builderChain),
				// Minimal assets catalog for utils/assets
				getAssetsObject: vi.fn((_chain: string) => ({
					nativeAssets: [{ symbol: 'KSM' }, { symbol: 'DOT' }],
					otherAssets: [],
				})),
			}
		})

		// Capture the transaction handed to signAndSend and return an unsubscribe fn
		vi.mock('@/utils/tx', () => ({
			signAndSend: vi.fn().mockImplementation(async ({ transaction }) => {
				// Return an unsubscribe function (no-op)
				return () => {}
			}),
		}))

		// Provide a minimal PolkadotApi mock; client is unused by our builder mock
		const papi = {
			getInstance: vi.fn().mockReturnValue({ client: {} }),
		} as any

		it('builds a serialized transfer tx that matches snapshot', async () => {
			const balanceService = {
				getBalances: vi.fn().mockResolvedValue([
					{ chain: Chains.Kusama, transferable: 1_000_000n },
					{ chain: Chains.AssetHubKusama, transferable: 200n },
				]),
			} as any

			const bridge = new XCMBridge(
				{
					chains: [Chains.Kusama, Chains.AssetHubKusama],
					getSigner: async () => ({}) as any,
					bridgeProtocols: ['XCM'],
				} as any,
				balanceService,
				papi,
			)

			const { signAndSend } = await import('@/utils/tx')

			await bridge.transfer(
				{
					amount: 123_456n,
					from: Chains.Kusama,
					to: Chains.AssetHubKusama,
					address: SUBSTRATE_ADDRESS,
					asset: Assets.KSM,
				},
				vi.fn(),
			)

			expect(signAndSend).toHaveBeenCalled()
			const callArg = (signAndSend as any).mock.calls.at(-1)?.[0]
			expect(callArg).toBeDefined()
			expect(callArg.transaction.decodedCall).toMatchSnapshot()
		}, 30000)
	},
)
