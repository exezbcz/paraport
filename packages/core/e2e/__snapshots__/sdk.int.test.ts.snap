// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration: XCMBridge transfer (serialized tx snapshot) > builds a serialized transfer tx that matches snapshot 1`] = `
{
  "method": "transfer_assets",
  "module": "PolkadotXcm",
  "parameters": {
    "assets": "...omitted for brevity...",
    "dest": "...omitted for brevity...",
  },
}
`;
